File: 2025/python/day_8/part_2.py
import sys
from .common import (
    FILE_PATH,
    Point,
    Segment,
    parse_line_into_point,
    parse_points_into_segments,
    sort_segments_by_length,
    island_sizes_p2,
)


def connect_all_pairs(segments: list[Segment], points: list[Point]) -> int:
    for i in range(len(segments)):
        segment = segments[i]
        segment.start.connect(segment.end)

        islands = island_sizes_p2(points)

        if len(islands) == 1:
            return segment.start.x * segment.end.x

    return -1


def day8_part2() -> None:
    try:
        running_id = 0
        points = []

        with open(FILE_PATH, "r") as f:
            for line in f:
                line = line.strip()
                point = parse_line_into_point(line, running_id)
                points.append(point)
                running_id += 1

        segments = sort_segments_by_length(parse_points_into_segments(points))
        solution = connect_all_pairs(segments, points)

        print(f"Solution: {solution}")

    except FileNotFoundError:
        print("File not found")
        sys.exit(1)
    except ValueError as e:
        print(f"Parsing error: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: 2025/ts/src/day-8/part-2.ts
import * as fs from 'fs';
import * as readline from 'readline';
import {
  PATH,
  Point,
  Segment,
  parseLineIntoPoint,
  parsePointsIntoSegments,
  sortSegmentsByLength,
  islandSizesP2,
} from './common';

function connectAllPairs(segments: Segment[], points: Point[]): number {
  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    segment.start.connect(segment.end);

    const islands = islandSizesP2(points);

    if (islands.length === 1) {
      return segment.start.x * segment.end.x;
    }
  }

  return -1;
}

export async function day8Part2(): Promise<void> {
  try {
    const fileStream = fs.createReadStream(PATH);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    let runningId = 0;
    const points: Point[] = [];

    for await (const line of rl) {
      const trimmedLine = line.trim();
      const point = parseLineIntoPoint(trimmedLine, runningId);
      points.push(point);
      runningId++;
    }

    const segments = sortSegmentsByLength(parsePointsIntoSegments(points));
    const solution = connectAllPairs(segments, points);

    console.log(`Solution: ${solution}`);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}
